stdlib/src/utils/static_tuple.mojo

```diff
@@ -240,3 +240,162 @@ struct StaticTuple[element_type: AnyRegType, size: Int](Sized):
         )
         Pointer(ptr).store(val)
         self = tmp
+
+
+# ===----------------------------------------------------------------------===#
+# Array
+# ===----------------------------------------------------------------------===#
+
+
+@value
+struct InlineArray[ElementType: CollectionElement, size: Int](Sized):
+    """A fixed-size sequence of size homogenous elements where size is a constant expression.
+
+    Parameters:
+        ElementType: The type of the elements in the array.
+        size: The size of the array.
+    """
+
+    alias type = __mlir_type[
+        `!pop.array<`, size.value, `, `, Self.ElementType, `>`
+    ]
+    var _array: Self.type
+    """The underlying storage for the array."""
+
+    # This constructor will always cause a compile time error if used.
+    # It is used to steer users away from uninitialized memory.
+    @always_inline
+    fn __init__(inout self):
+        constrained[
+            False,
+            (
+                "Initialize with either a variadic list of arguments or a"
+                " default fill element."
+            ),
+        ]()
+        self._array = __mlir_op.`kgen.undef`[_type = Self.type]()
+
+    @always_inline
+    fn __init__(inout self, fill: Self.ElementType):
```

Comment:
bikeshed: is `fill` the right word here?

Reply:
`SIMD` just uses `value` for this kind of constructor. Not sure if that's a great precedent or not, although it's probably clear enough. From the name I wouldn't be sure if `fill` was a value or a boolean. Maybe `fill_value` or `fill_with`?

Also, have you tried to create an `InlineArray` of size 1? It looks like the two constructors would be ambiguous in that case. Making  this a keyword-only argument would avoid that potential issue.

`InlineArray[Int, 10](fill_with=0xFF)`

Reply:
At least on my machine, the constructor of size 1 seems to work fine.

Reply:
They're not ambiguous as @lsh points out.  Consider the following smaller program that shows this is the case:

```
struct S:
    fn __init__(inout self, value: Int):
        print("single argument constructor")

    fn __init__(inout self, *values: Int):
        print("variadic argument constructor")

fn main():
    var s1 = S(42) # prints single argument constructor
    var s2 = S(42, 100) # prints variadic argument constructor
```

Re the name, I'm fine with `fill`, `default_value`, `value`, or any of them.  I don't feel strongly and we can easily change it at any time since this isn't a keyword-argument-only constructor (therefore it's just an implementation detail at this point).